### PR TITLE
[AGENT-4244] Extend Batch Scoring operator to support credential_id input from XCom

### PR DESCRIPTION
**Pull request by oleksandr-saienko**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Extended ScorePredictionsOperator to support credentials_id input from preconfigured credentials, added example dags for S3, Azure Blob Storage and GCP Storage

## Rationale
Users should be able to use ScorePredictionsOperator with preconfigured credentials to simplify DAG creation process

### CODE REVIEW

**Comment by oleksandr-saienko** [comment]
Jenkins run

**Comment by oleksandr-saienko** [comment]
Jenkins run

File: README.md [hunk 1]
```diff
@@ -175,15 +175,33 @@ in the `context["params"]` variable, e.g. getting a training data you would use
     Scores batch predictions against the deployment.
 
     Prerequisites:
-    - [S3 credentials added to DataRobot via Python API client](https://datarobot-public-api-client.readthedocs-hosted.com/en/latest-release/reference/admin/credentials.html#s3-credentials).
-      You need the `creds.credential_id` for the `credential_id` parameter in the config.
+    - You can use GetCredentialIdOperator to pass credential_id from preconfigured DataRobot Credentials (Airflow Connections)
+    - or you can manually set `credential_id` parameter in the config. [S3 credentials added to DataRobot via Python API client](https://datarobot-public-api-client.readthedocs-hosted.com/en/latest-release/reference/admin/credentials.html#s3-credentials).
```
Comment (elatt) [inline README.md:179]:
```suggestion
    - You can use GetCredentialIdOperator to pass `credential_id` from preconfigured DataRobot Credentials (Airflow Connections)
      or you can manually set `credential_id` parameter in the config. [S3 credentials added to DataRobot via Python API client](https://datarobot-public-api-client.readthedocs-hosted.com/en/latest-release/reference/admin/credentials.html#s3-credentials).
```

File: datarobot_provider/example_dags/datarobot_aws_s3_batch_scoring_dag.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+# All rights reserved.
+#
+# This is proprietary source code of DataRobot, Inc. and its affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+"""
+Example of Aiflow DAG for DataRobot Batch Scoring using S3 as source and destination,
+using a preconfigured "DataRobot AWS Credentials" from Airflow Connections.
+DataRobot AWS Credentials can be configured using Airflow UI (Admin->Connections) or Airflow API
+Config example for this dag:
```
Comment (elatt) [inline datarobot_provider/example_dags/datarobot_aws_s3_batch_scoring_dag.py:12]:
```suggestion
Example of Aiflow DAG for DataRobot Batch Scoring using S3 as source and destination,
and using a preconfigured "DataRobot AWS Credentials" from Airflow Connections.
DataRobot AWS Credentials can be configured using Airflow UI (Admin->Connections) or Airflow API
Config example for this dag:
```

File: datarobot_provider/example_dags/datarobot_azure_storage_batch_scoring_dag.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+# All rights reserved.
+#
+# This is proprietary source code of DataRobot, Inc. and its affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+"""
+Example of Aiflow DAG for DataRobot Batch Scoring using Azure Blob Storage as source and destination,
+using a preconfigured "DataRobot Azure Storage Credentials" from Airflow Connection.
+DataRobot Azure Storage Credentials can be configured using Airflow UI (Admin->Connections) or Airflow API
+Config example for this dag:
```
Comment (elatt) [inline datarobot_provider/example_dags/datarobot_azure_storage_batch_scoring_dag.py:12]:
```suggestion
Example of Aiflow DAG for DataRobot Batch Scoring using Azure Blob Storage as source and destination,
and using a preconfigured "DataRobot Azure Storage Credentials" from Airflow Connection.
DataRobot Azure Storage Credentials can be configured using Airflow UI (Admin->Connections) or Airflow API
Config example for this dag:
```

File: datarobot_provider/example_dags/datarobot_azure_storage_batch_scoring_dag.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                "url": "https:// ... .blob.core.windows.net/.../input_filename.csv",
+            },
+            "output_settings": {
+                "type": "azure",
+                "url": "https:// ... .blob.core.windows.net/.../output_filename.csv",
+            },
+        },
+    },
+)
+def datarobot_azure_storage_batch_scoring(deployment_id="put_your_deployment_id"):
```
Comment (elatt) [inline datarobot_provider/example_dags/datarobot_azure_storage_batch_scoring_dag.py:58]:
```suggestion
def datarobot_azure_storage_batch_scoring(deployment_id=None):
```
ditto about `None` here?

File: datarobot_provider/example_dags/datarobot_gcp_storage_batch_scoring_dag.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+# Copyright 2022 DataRobot, Inc. and its affiliates.
+#
+# All rights reserved.
+#
+# This is proprietary source code of DataRobot, Inc. and its affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+"""
+Example of Aiflow DAG for DataRobot Batch Scoring using Google Cloud Storage as source and destination,
+using a preconfigured "DataRobot GCP Credentials" from Airflow Connection.
```
Comment (elatt) [inline datarobot_provider/example_dags/datarobot_gcp_storage_batch_scoring_dag.py:10]:
```suggestion
and using a preconfigured "DataRobot GCP Credentials" from Airflow Connection.
```

File: datarobot_provider/example_dags/datarobot_gcp_storage_batch_scoring_dag.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            "output_settings": {
+                "type": "gcp",
+                "url": "gs://bucket_name/output_file_name.csv",
+            },
+            # If passthrough columns are required, use this line:
+            "passthrough_columns": ['column1', 'column2'],
+        },
+    },
+)
+def datarobot_gcp_batch_scoring(deployment_id="put_your_deployment_id"):
```
Comment (elatt) [inline datarobot_provider/example_dags/datarobot_gcp_storage_batch_scoring_dag.py:60]:
```suggestion
def datarobot_gcp_batch_scoring(deployment_id=None):
```

File: tests/unit/dags/conftest.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+
+import pytest
+from airflow.models import DagBag
+
+
+@pytest.fixture()
+def dagbag(provider_dir):
+    return DagBag(dag_folder=f"{str(provider_dir)}/example_dags", include_examples=False)
```
Comment (elatt) [inline tests/unit/dags/conftest.py:15]: Nice refactoring 👍

File: datarobot_provider/example_dags/datarobot_aws_s3_batch_scoring_dag.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            "output_settings": {
+                "type": "s3",
+                "url": "s3://bucket_name/output_file_name.csv",
+            },
+        },
+    },
+)
+def datarobot_s3_batch_scoring(deployment_id="put_your_deployment_id"):
+    if not deployment_id:
+        raise ValueError("Invalid or missing `deployment_id` value")
```
[the remark this answers is not in the document]
Comment (oleksandr-saienko) [reply datarobot_provider/example_dags/datarobot_aws_s3_batch_scoring_dag.py:60]: Agree that should be None

File: tests/unit/dags/predictions/test_datarobot_azure_storage_batch_scoring_dag.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+# All rights reserved.
+#
+# This is proprietary source code of DataRobot, Inc. and its affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+
+from datarobot_provider.example_dags.datarobot_azure_storage_batch_scoring_dag import (
+    datarobot_azure_storage_batch_scoring,
+)
+from tests.unit.dags.conftest import assert_dag_dict_equal
```
Comment (elatt) [inline tests/unit/dags/predictions/test_datarobot_azure_storage_batch_scoring_dag.py:12]: Don't do this.

File: tests/unit/dags/conftest.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def assert_dag_dict_equal(source, dag):
+    assert dag.task_dict.keys() == source.keys()
+    for task_id, downstream_list in source.items():
+        assert dag.has_task(task_id)
+        task = dag.get_task(task_id)
+        assert task.downstream_task_ids == set(downstream_list)
```
Comment (elatt) [inline tests/unit/dags/conftest.py:23]: The common way to have these helper functions is either use the pytest-plugin: https://pypi.org/project/pytest-helpers-namespace/

or wrap this code in a class (as a static method), e.g.
```python
class _DagAssertions:
    @staticmethod
    def assert_dict_equal(source, dag):
        assert dag.task_dict.keys() == source.keys()
        for task_id, downstream_list in source.items():
            assert dag.has_task(task_id)
            task = dag.get_task(task_id)
            assert task.downstream_task_ids == set(downstream_list)

@pytest.fixture
def DagAssertions():
    return _DagAssertions
```

File: tests/unit/dags/predictions/test_datarobot_azure_storage_batch_scoring_dag.py (hunk 9 above)
Comment (elatt) [reply tests/unit/dags/predictions/test_datarobot_azure_storage_batch_scoring_dag.py:12]:
```suggestion
```

**Review by elatt — COMMENTED** [review]
Looks good other than the issue I noted with `conftest.py`. I realize you added `__init__.py` files into the test directories but technically that isn't the standard way of dealing with pytest and [how it imports test modules](https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html). I don't think we should rely on the modules (or `conftest.py`) being importable because it could paint us into a corner and there are other canonical ways to share a helper function like you are doing in pytest.

**Review by elatt — APPROVED** [review]
LGTM

**Review by dulcardo — APPROVED** [review]
LGTM

**Pull request merged**